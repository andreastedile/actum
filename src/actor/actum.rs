use crate::core::actor_cell::ActorCell;
use crate::core::actor_ref::ActorRef;
use crate::core::actor_task::ActorTask;
use crate::core::actor_to_spawn::CreateActorResult;
use crate::core::message_receiver::MessageReceiver;
use futures::channel::mpsc;

/// Instantiates the actor tree hierarchy.
///
/// The argument of this function is the root actor of the actor tree hierarchy.
/// From within the root actor, you can [create](prelude::CreateChild::create_child) new child actors.
/// Therefore, this function effectively instantiates the whole actor tree hierarchy.
///
/// Returns a struct containing the [ActorRef] of the root actor (generic over parameter `M`) and a special control structure.
/// The control structure has a method that returns a future which runs the root actor, joins its child actors (if any), and resolves with the value returned by the root actor (generic over parameter `Ret`).
/// Once this future resolves, the entire actor tree hierarchy has returned.
/// This future is **runtime agnostic**.
///
/// Note that you can send messages to the root actor before awaiting the future.
///
/// # Examples
///
/// You can pass a closure. In this way, you can also pass values to the actor.
///
/// ```rust
/// use actum::prelude::*;
///
/// #[tokio::main]
/// async fn main() {
///     let numbers = vec![1, 2, 3]; // gets moved into the closure
///     let CreateActorResult { task, mut actor_ref } = actum::<String, _, _, u64>(|cell, mut receiver, _me| async move {
///         let m = receiver.recv().await.into_message().unwrap();
///         println!("received: {}", m);
///
///         let sum = numbers.iter().sum();
///         (cell, sum)
///     });
///     actor_ref.try_send("hello from main!".to_string()).unwrap();
///     let sum = task.run_task().await;
///     println!("sum = {}", sum);
/// }
/// ```
///
/// You can pass a function pointer.
///
/// ```rust
/// use actum::prelude::*;
///
/// async fn root_actor<C, R>(cell: C, mut receiver: R, _me: ActorRef<u64>) -> (C, u64)
/// where
///     C: CreateChild,
///     R: ReceiveMessage<u64>,
/// {
///     let mut sum = 0;
///     for _ in 0..3 {
///         let m = receiver.recv().await.into_message().unwrap();
///         println!("received: {}", m);
///         sum += m;
///     }
///     (cell, sum)
/// }
///
/// #[tokio::main]
/// async fn main() {
///     let CreateActorResult { task, mut actor_ref } = actum(root_actor);
///     actor_ref.try_send(1).unwrap();
///     actor_ref.try_send(2).unwrap();
///     actor_ref.try_send(3).unwrap();
///     let sum = task.run_task().await;
///     println!("sum = {}", sum);
/// }
/// ```
pub fn actum<M, F, Fut, Ret>(f: F) -> CreateActorResult<M, ActorTask<M, F, Fut, Ret, (), (), ()>>
where
    M: Send + 'static,
    F: FnOnce(ActorCell<()>, MessageReceiver<M, ()>, ActorRef<M>) -> Fut + Send + 'static,
    Fut: Future<Output = (ActorCell<()>, Ret)> + Send + 'static,
    Ret: Send + 'static,
{
    let m_channel = mpsc::channel::<M>(100);
    let actor_ref = ActorRef::new(m_channel.0);
    let receiver = MessageReceiver::new(m_channel.1, ());

    let cell = ActorCell::new(());

    let task = ActorTask::new(f, cell, receiver, actor_ref.clone(), (), None);

    CreateActorResult::new(task, actor_ref)
}
