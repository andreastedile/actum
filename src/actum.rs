use crate::actor_cell::ActorCell;
use crate::actor_path::ActorPath;
use crate::actor_ref::ActorRef;
use crate::actor_system::{ActorSystem, NameError};
use futures::FutureExt;
use std::any::Any;
use std::future::Future;
use std::panic::AssertUnwindSafe;
use tokio::sync::mpsc;
use tokio_util::sync::CancellationToken;
use tracing::{instrument, trace_span, Instrument};

#[derive(Debug)]
pub enum ActumError {
    Name(NameError),
    Panic(Box<dyn Any + Send>),
}

#[instrument(level = "trace", skip(actor_fn), ret)]
pub fn actum<M, Fut>(
    name: &str,
    actor_fn: impl FnOnce(ActorCell<M>) -> Fut,
) -> Result<(), ActumError>
where
    M: Send + 'static,
    Fut: Future + Send + 'static,
    Fut::Output: Send + 'static,
{
    let runtime = tokio::runtime::Runtime::new().unwrap();

    let system = match ActorSystem::new(name, runtime.handle().clone()) {
        Ok(system) => system,
        Err(error) => return Err(ActumError::Name(error)),
    };

    let path = ActorPath::new("guardian");

    let (message_sender, message_receiver) = mpsc::channel::<M>(42);
    let me = ActorRef::<M>::new(path.clone(), message_sender);

    let (_stop_sender, stop_receiver) = mpsc::unbounded_channel::<()>();

    let drop = CancellationToken::new();

    let cell = ActorCell::<M>::new(system, me, message_receiver, stop_receiver, drop.clone());

    let future = actor_fn(cell);

    let span = trace_span!(parent: None, "actor", path = ?path);
    let panic = runtime.block_on(
        async move {
            let panic: Option<Box<dyn Any + Send>> =
                AssertUnwindSafe(future).catch_unwind().await.err();

            drop.cancelled_owned().await;

            panic
        }
        .instrument(span),
    );

    panic.map_or(Ok(()), |panic| Err(ActumError::Panic(panic)))
}
