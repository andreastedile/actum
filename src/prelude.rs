pub use crate::actor::Actor;
pub use crate::actor_bounds::ActorBounds;
pub use crate::actor_cell::actor_task::RunTask;
pub use crate::actor_cell::Stop;
pub use crate::actor_ref::ActorRef;
pub use crate::actum;
