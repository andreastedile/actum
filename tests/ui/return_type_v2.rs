use actum::prelude::*;

async fn root_v2<A, R>(cell: A, mut receiver: R, _me: ActorRef<u64>) -> (A, u32)
where
    A: Actor<u64, String>,
    //            ^^^^^^ it complies, but it should not: String != u32
    R: Receiver<u64>,
{
    let _ = receiver.recv().await;
    (cell, 4)
}

#[tokio::main]
async fn main() {
    let _root = actum(root_v2);
}
