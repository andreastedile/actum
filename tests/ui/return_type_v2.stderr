error[E0277]: the trait bound `ActorCell<(), u32>: actum::actor::Actor<u64, String>` is not satisfied
  --> tests/ui/return_type_v2.rs:15:23
   |
15 |     let _root = actum(root_v2);
   |                       ^^^^^^^ the trait `actum::actor::Actor<u64, String>` is not implemented for `ActorCell<(), u32>`
   |
   = help: the following other types implement trait `actum::actor::Actor<M, Ret>`:
             ActorCell<(), Ret>
             ActorCell<TestExtension<Ret>, Ret>
note: required by a bound in `root_v2`
  --> tests/ui/return_type_v2.rs:5:8
   |
3  | async fn root_v2<A, R>(cell: A, mut receiver: R, _me: ActorRef<u64>) -> (A, u32)
   |          ------- required by a bound in this function
4  | where
5  |     A: Actor<u64, String>,
   |        ^^^^^^^^^^^^^^^^^^ required by this bound in `root_v2`

error[E0277]: the trait bound `ActorCell<(), u32>: actum::actor::Actor<u64, String>` is not satisfied
  --> tests/ui/return_type_v2.rs:15:17
   |
15 |     let _root = actum(root_v2);
   |                 ^^^^^ the trait `actum::actor::Actor<u64, String>` is not implemented for `ActorCell<(), u32>`
   |
   = help: the following other types implement trait `actum::actor::Actor<M, Ret>`:
             ActorCell<(), Ret>
             ActorCell<TestExtension<Ret>, Ret>
note: required by a bound in `root_v2`
  --> tests/ui/return_type_v2.rs:5:8
   |
3  | async fn root_v2<A, R>(cell: A, mut receiver: R, _me: ActorRef<u64>) -> (A, u32)
   |          ------- required by a bound in this function
4  | where
5  |     A: Actor<u64, String>,
   |        ^^^^^^^^^^^^^^^^^^ required by this bound in `root_v2`

error[E0277]: the trait bound `ActorCell<(), u32>: actum::actor::Actor<u64, String>` is not satisfied
  --> tests/ui/return_type_v2.rs:15:17
   |
15 |     let _root = actum(root_v2);
   |                 ^^^^^^^^^^^^^^ the trait `actum::actor::Actor<u64, String>` is not implemented for `ActorCell<(), u32>`
   |
   = help: the following other types implement trait `actum::actor::Actor<M, Ret>`:
             ActorCell<(), Ret>
             ActorCell<TestExtension<Ret>, Ret>
note: required by a bound in `root_v2`
  --> tests/ui/return_type_v2.rs:5:8
   |
3  | async fn root_v2<A, R>(cell: A, mut receiver: R, _me: ActorRef<u64>) -> (A, u32)
   |          ------- required by a bound in this function
4  | where
5  |     A: Actor<u64, String>,
   |        ^^^^^^^^^^^^^^^^^^ required by this bound in `root_v2`
